### DISCUSSION TITLE
Livewire request resets config value.

### DISCUSSION THREAD

**Discussion by Mo7sin** [livewire/livewire#5739] [1. Help]
Hello, I need help understanding what is wrong with my case.

Livewire version: 2.12.3
Laravel version: 10.13.5
alpinejs@3.11.1
@alpinejs/intersect@3.11.1

Let me explain my situation,

I have one standard HTTP request which shows a single Book. **https://books.test/books/123**

On the same page, I have several livewire components which get initialized when they become visible in the viewport (using alpinejs intersect plugin)

```
<div x-intersect.once="@this.call('fetch');"
         x-transition.duration.1000ms>
        //fetched items here.
</div>
```

in **AppServiceProvider** I get the country iso code from the **request()->server('HTTP_CF_IPCOUNTRY')**,  all other mentioned requests (Standard HTTP, and livewire components fetch requests) rely on that value.

So far so good.

The issue is happening when I get the country code from the Query string **https://books.test/books/123?country=US**

to do that I added a new key with a default value **"NULL"** in my **app.php**  called **country** and registered a new middleware 

```
<?php

namespace App\Http\Middleware;

use Closure;
use Illuminate\Support\Facades\Config;

class SetCountryMiddleware
{
    public function handle($request, Closure $next)
    {
        $country = $request->query('country');

        if ($country) { // checking if the query string has the country key before updating the value.
            Config::set('app.country', $country);
        }

        return $next($request);
    }
}
```

```
app/Http/Kernel.php

protected $middleware = [
    // Other middleware...
    \App\Http\Middleware\SetCountryMiddleware::class,
];
```

for the standard HTTP request, the value was updated correctly, and everything worked as expected.

Now its livewire components turn, 

As we know, before livewire message request is finished it has to pass all middlewares defined (just like a standard request).

and here is exactly the issue, livewire request doesn't read from the query string, and even weirder it resets that config country to its default value **NULL** after the request has finished.

I've tried to store country from the query string in the session instead of config, but still, livewire cannot read that session value and return **NULL**, after the request has finished.

```
<?php

namespace App\Http\Middleware;

use Closure;
use Illuminate\Support\Facades\Config;

class SetCountryMiddleware
{
    public function handle($request, Closure $next)
    {
        $country = $request->query('country');

        if ($country) { // checking if the query string has the country key before updating the value.
            session(['app.country', $country]);
        }

        return $next($request);
    }
}
```

So, has anyone got any issue like that? is it something wrong I'm doing? Any advice?

Thank you.

**Comment by lrljoe** [comment]
You need to add your middleware into your LW for persistence across multiple requests.

Session would be the right approach for this though 

What's the order of your middleware in full?

Where in your LW component are you trying to access the session variable?  Is there a reason you don't use the querystring approach from Livewire and persist once set?

**Comment by Mo7sin** [comment]
Hi @lrljoe 

> You need to add your middleware into your LW for persistence across multiple requests.

Where should middleware be added?

I have in config/livewire.php

`'middleware_group' => 'web',`

So I moved my middleware from `$middleware` to the 'web' group in `kernel.php` and I placed it as the last item in the array

```
    protected $middlewareGroups
        = [
            'web' => [
                EncryptCookies::class,
                AddQueuedCookiesToResponse::class,
                StartSession::class,
                ShareErrorsFromSession::class,
                VerifyCsrfToken::class,
                LoadCommonData::class,
                LastSeen::class,
                SubstituteBindings::class,
                SetCountryMiddleware::class,
            ],
   ],
.....
```

And I did some debugging to make sure middleware is executed, it does get executed thou but the value is still NULL for `config('app.country')`.

----------------------
> Where in your LW component are you trying to access the session variable? Is there a reason you don't use the querystring approach from Livewire and persist once set?

I'm not trying to access session or queryString from within livewire component itself, The middleware is updating `config('app.country')` value, So when livewire component gets visible and the request is initialized it should read that updated value from `AppServiceProvider` as it relies on `config('app.country')` for every request to determine the currency and stuff,
but rather livewire request reads the default value from `app/config.php` which is `NULL`.

**Reply by lrljoe** [reply]
Persisting Middlware for LW Requests is in your AppServiceProvider
```php
Livewire::addPersistentMiddleware([
    YourOwnMiddleware::class,
]);
```

If you use the Log:: function in your middleware, is it definitely setting the config value?

Have you tried this approach instead:
```
config(['app.country' => $request->query('country')]);
```

I would definitely be using the session value to store something related to the current session though!  I'd also suggest that you set a default value for app.country.

Ideally you should use the localisation approach if possible to allow for translation of strings as well as currencies etc.

**Reply by Mo7sin** [reply]
I tried the session approach now and moved the whole process to the middleware instead of `AppServiceProvider` and it seems to be working now. the only thing that is not clear to me is why the `query string` is always empty in the livewire request.

So, if I rely on `$request->query('country')` it will never work since it is always null, So, for now, I get the country value from `$request->server('HTTP_REFERER')`  instead.

**Reply by lrljoe** [reply]
How are you using the querystring in Livewire? Can you post what you're trying?

The following in your Livewire component will keep your $this->country synchronised with the value of country from the query string 

```php

public string $country;
 
protected $queryString = ['country'];

```

**Reply by Mo7sin** [reply]
The livewire component doesn't use querystring internally; instead, it renders another sub-component that relies on that `country` value I'm trying to update from HTTP standard query string value.

After I moved the whole process to a middleware, it's working fine (with small issues that I still working on).

Thank you @lrljoe for your participation.
